python/tank_vendor/shotgun_deploy/configuration.py

```diff
@@ -0,0 +1,965 @@
+                                   should be set to None.
+        :param namespace: name space string, typically one short word,
+                          e.g. 'maya', 'rv', 'desktop'.
+        :param bundle_cache_fallback_paths: List of additional paths where apps are cached.
+        """
+        self._sg_connection = sg
+        self._descriptor = descriptor
+        self._project_id = project_id
+        self._pipeline_config_id = pipeline_config_id
+        self._namespace = namespace
+        self._bundle_cache_fallback_paths = bundle_cache_fallback_paths
+
+    def __repr__(self):
+        return "<Config with id %s, project id %s and base %s>" % (
+            self._pipeline_config_id,
+            self._project_id,
+            self._descriptor
+        )
+
+    def get_path(self, platform=sys.platform):
+        """
+        Returns the path to the installed configuration
+        on the given os. Note how the location returned by
+        this method is the root folder of an installed configuration.
+
+        This root folder in turn contains a 'config' folder which contains
+        the configuration described by the config descriptor.
+
+        :param platform: Operating system platform
+        :return: path on disk as string
+        """
+        raise NotImplementedError
+
+    def status(self):
+        """
+        Compares the actual configuration installed on disk against the
+        associated configuration described by the descriptor passed in via
+        the class constructor.
+
+        :returns: LOCAL_CFG_UP_TO_DATE, LOCAL_CFG_MISSING,
+                  LOCAL_CFG_OLD, or LOCAL_CFG_INVALID
+        """
+        raise NotImplementedError
+
+    def update_configuration(self):
+        """
+        Ensure that the configuration is up to date with the one
+        given by the associated descriptor.
+
+        This method fails gracefully and attempts to roll back to a
+        stable state on failure.
+        """
+        raise NotImplementedError
+
+    def get_tk_instance(self, sg_user):
+        """
+        Returns a tk instance for this configuration.
+
+        :param sg_user: Authenticated Shotgun user to associate
+                        the tk instance with.
+        """
+        # @todo - check if there is already a tank instance, in that case unload or warn
+        path = self.get_path()
+        core_path = os.path.join(path, "install", "core", "python")
+        sys.path.insert(0, core_path)
+        import tank
+        tank.set_authenticated_user(sg_user)
+        tk = tank.tank_from_path(path)
+        log.info("API created: %s" % tk)
+        return tk
+
+    def _ensure_project_scaffold(self):
+        """
+        Creates all the necessary files on disk for a basic config scaffold.
+
+        - Sets up basic folder structure for a config
+        - Copies the configuration into place.
+
+        :returns: True if a scaffold was created, False if a complete or
```

Comment:
Maybe I'm reading this wrong, but the way the code looks, it looks like `config_exists` returns `True` if the path already existed, `False` otherwise.